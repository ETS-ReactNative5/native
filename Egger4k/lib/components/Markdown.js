import React from "react";
import {StyleSheet} from 'react-native';
import Renderer from 'react-native-markdown-renderer'
import {theme} from "../../theme";

export default function Markdown(props){
    return <Renderer style={styles}>{props.children}</Renderer>
}

const styles = StyleSheet.create({
    text: {
        fontSize: 18
    },
    heading1:{
        fontFamily: theme.fontFamilyH1,
        color: theme.brandPrimary,
        fontSize: 28,
    },
    heading2:{
        fontFamily: theme.fontFamilyH2,
        color: theme.brandSecondary,
        fontSize: 24,
    }
});