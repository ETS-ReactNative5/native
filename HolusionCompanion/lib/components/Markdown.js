import React, {useContext} from "react";
import {StyleSheet} from 'react-native';

import ObjectLink from "./ObjectLink";

import Renderer, { MarkdownIt } from 'react-native-markdown-display';

import { ThemeContext } from "./style";

const BAD_PROTO_RE = /^(vbscript|javascript|data):/;
const GOOD_DATA_RE = /^data:image\/(gif|png|jpeg|webp);/;

function useTheme(){
  const theme = useContext(ThemeContext);
  return {mdTheme: {
    heading1:{
      fontSize: theme.fontSize.h1,
      //fontFamily: theme.fontFamily.h2,
      color: theme.colors.primary,
    },
    heading2:{
      fontSize: theme.fontSize.h2,
      //fontFamily: theme.fontFamily.h2,
      color: theme.colors.secondary,
    },
    text: {
      fontSize: theme.fontSize.default,
    },
  }}
}

export default function Markdown(props){
  const mdTheme = useTheme();
  const parser = MarkdownIt({
    typographer: true,
    validateLink: (url)=> {
      str = url.trim().toLowerCase();
      return BAD_PROTO_RE.test(str) ? (GOOD_DATA_RE.test(str) ? true : false) : true;
    }
  });

  const rules = {
    link: (node, children, parent, styles) => (
      <ObjectLink to={node.attributes.href}
        key={node.key}
        style={styles.link}>
        {children}
      </ObjectLink>
    ),
    blocklink: (node, children, parent, styles) => (
      <ObjectLink to={node.attributes.href}
        key={node.key}
        style={styles.blocklink}>
        <View style={styles.image}>{children}</View>
      </ObjectLink>
    ),
  }

  const styles = Array.isArray(props.style)? props.style.reduce((acc, v)=>{return {...acc, ...v}}, {}): props.style;
  return <Renderer 
    rules={rules}
    markdownit={parser}
    allowedImageHandlers={["file://", 'data:image/png;base64', 'data:image/gif;base64', 'data:image/jpeg;base64', "http://", "https://"]} 
    defaultImageHandler={null}
    style={mdTheme}
  >
    {props.children}
  </Renderer>
  
}



//export default connectStyle('Holusion.Markdown', mdTheme)(Markdown);
