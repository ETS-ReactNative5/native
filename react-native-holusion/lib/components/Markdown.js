import React from "react";
import {StyleSheet} from 'react-native';
import {connectStyle} from "native-base";
import {Link} from "@react-navigation/native";

import Renderer, { MarkdownIt } from 'react-native-markdown-display'

const BAD_PROTO_RE = /^(vbscript|javascript|data):/;
const GOOD_DATA_RE = /^data:image\/(gif|png|jpeg|webp);/;


class Markdown extends React.Component {
  constructor(props){
      super(props);
      this.parser = MarkdownIt({typographer: true});
      this.parser.validateLink = (url)=> {
        str = url.trim().toLowerCase();
        return BAD_PROTO_RE.test(str) ? (GOOD_DATA_RE.test(str) ? true : false) : true;
      };
  }
  static handlers = ["file://", 'data:image/png;base64', 'data:image/gif;base64', 'data:image/jpeg;base64', "http://", "https://"]
  rules = {
    link: (node, children, parent, styles) => (
      <Link to={`/${node.attributes.href}`}
        key={node.key}
        style={styles.link}>
        {children}
      </Link>
    ),
    blocklink: (node, children, parent, styles) => (
      <Link to={`/${node.attributes.href}`}
        key={node.key}
        style={styles.blocklink}>
        <View style={styles.image}>{children}</View>
      </Link>
    ),
  }

  render(){
    //console.log("Render markdown : ", this.props.children);
    const styles = Array.isArray(this.props.style)? this.props.style.reduce((acc, v)=>{return {...acc, ...v}}, {}): this.props.style;
    return <Renderer 
      rules={this.rules}
      markdownit={this.parser}
      allowedImageHandlers={["file://", 'data:image/png;base64', 'data:image/gif;base64', 'data:image/jpeg;base64', "http://", "https://"]} 
      defaultImageHandler={null}
      style={styles}
    >
      {this.props.children}
    </Renderer>
  }
}

const mdTheme = StyleSheet.create({

});



export default connectStyle('Holusion.Markdown', mdTheme)(Markdown);
