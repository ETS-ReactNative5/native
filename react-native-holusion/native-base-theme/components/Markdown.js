import variable from './../variables/platform';

export default (variables /* : * */ = variable) => {
    const mdTheme = {
        heading1:{
            fontSize: variables. fontSizeH1,
            fontFamily: variables.fontFamilyH1,
            color: variables.brandPrimary,
        },
        heading2:{
            fontSize: variables. fontSizeH2,
            fontFamily: variables.fontFamilyH2,
            color: variables.brandSecondary,
        },
        text: {
            fontSize: variables.DefaultFontSize,
        },
    };

  return mdTheme;
};

