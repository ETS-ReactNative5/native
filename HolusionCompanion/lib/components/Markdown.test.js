import React from "react";
import {Text} from "react-native";
import { render, fireEvent, act } from '@testing-library/react-native'

jest.mock("./ObjectLink",()=>(()=> null));

import Markdown from "./Markdown";


describe("<Markdown>", ()=>{
  test("handle images", ()=>{
    const res = render(<Markdown>![alt text](file://example.com/foo.png)</Markdown>)
    expect(res.toJSON()).toMatchSnapshot();
  })
  test("handle links", ()=>{
    const res = render(<Markdown>[text](foo)</Markdown>)
    expect(res.toJSON()).toMatchSnapshot();
  })
})