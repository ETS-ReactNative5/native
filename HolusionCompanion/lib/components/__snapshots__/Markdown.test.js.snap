// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Markdown> handle images 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
        "marginBottom": 10,
        "marginTop": 10,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "flexGrow": 1,
          },
          Object {
            "height": 0,
          },
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Image
        accessibilityLabel="alt text"
        accessible={true}
        indicator={true}
        onError={[Function]}
        onLayout={[Function]}
        onLoad={[Function]}
        onLoadStart={[Function]}
        source={
          Object {
            "uri": "file://path/to/foo.png",
          }
        }
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "height": undefined,
              "width": undefined,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`<Markdown> handle links 1`] = `
<View
  style={Object {}}
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
        "marginBottom": 10,
        "marginTop": 10,
        "width": "100%",
      }
    }
  >
    <Text
      style={Object {}}
    />
  </View>
</View>
`;
